### PR TITLE
Cli test login

### PR DESCRIPTION
**Pull request by dheles**
- [X] Have you run `make test` locally and ensured that all tests pass?
- [X] Have you run `make coverage` locally and ensured that coverage did not drop?

*If* you're changing something in the library:
- [ ] Have you introduced unit tests for your changes?

*If* you're changing something in the CLI:
- [ ] Have you updated the manual pages and shell completion (if necessary)?
- [X] Have you introduced CLI unit tests for your changes and ensured that `make test-cli` passes?

Adds cli tests for login
Adjusts login to return error when no record found

### CODE REVIEW

**Comment by dheles** [comment]
notes:
- added more failure conditions, as per previous suggestion
- now that TTY check does not impose itself, standard aruba syntax works*, so i used it here
- did not add `# frozen_string_literal: true`, as this seems to break aruba's method of providing input
- added test for session argument, rather than saving all session testing for `test_session.rb`. there are numerous commands that have a session argument and it seems a good idea to test all arguments (when possible) for each command in its respective test. (at the very least, it reduces the chance of forgetting to cover something)

*terse input commands (e.g. `kbsecret new login -x ...`) are the remaining exception

File: test/cli/test_login.rb [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    username = "user"
+    password = "pass"
+    output = <<~OUTPUT
+      Label: #{label}
+      \tUsername: #{username}
+      \tPassword: #{password}
+    OUTPUT
+
+    # create login:
+    run_command "kbsecret new login #{label}"
```
Comment (woodruffw) [inline test/cli/test_login.rb]: Let's use the block form of `run_command`, since it makes it easier to transpose test cases around and lets us distinguish between stopping a command and waiting for it to finish:

```ruby
run_command "kbsecret new login #{label}" do |cmd|
  cmd.stdin.puts username, password
  cmd.stdin.close
  cmd.wait
end
```

File: test/cli/test_login.rb [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    # remove login:
+    run_command_and_stop "kbsecret rm #{label}"
+  end
+
+  def test_terse_separator
+    label = "test-terse-separator"
+    username = "user"
+    password = "pass"
+    separator = "^"
+    output = "#{label}#{separator}#{username}#{separator}#{password}"
```
Comment (woodruffw) [inline test/cli/test_login.rb]: `[label, username, password].join separator`

File: test/cli/test_login.rb [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+  end
+
+  def test_all
+    session = "test-session"
+    label = "test-session-login"
+    label1 = "test-session-login1"
+    username = "user"
+    password = "pass"
+   # NOTE: need regex since we can't be sure of the order in which the logins will be retreived
+    pattern = "Label: (#{label}|#{label1})\n\tUsername: #{username}\n\tPassword: #{password}\nLabel: (#{label1}|#{label})\n\tUsername: #{username}\n\tPassword: #{password}\n"
```
Comment (woodruffw) [inline test/cli/test_login.rb]: Use regexp literals:

```ruby
pattern = /Label: .../
error_pattern = /No such record/
```

File: test/cli/test_login.rb [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    error_pattern = "No such record"
+
+    # create a new session
+    run_command_and_stop "kbsecret session new -r test #{session}"
+
+    # fail to retrieve nonexistent logins:
+    run_command "kbsecret login -a"
+    stop_all_commands
+    assert_match(/#{error_pattern}/, last_command_started.stderr.chomp)
```
Comment (woodruffw) [inline test/cli/test_login.rb]: Don't interpolate here or below, just use the Regexp object directly:

```ruby
assert_match error_pattern, whatever
```

File: test/cli/test_login.rb [hunk 5]
```diff
@@ -0,0 +1,237 @@
+require_relative "test_helper"
+require "aruba"
+
+class CLITest < Minitest::Test
```
Comment (woodruffw) [inline test/cli/test_login.rb]: Give this class a unique name, or we'll probably end up clobbering multiple methods across the test suite. Maybe something like `CLILoginTest`?

File: lib/kbsecret/cli/kbsecret-login [hunk 6]
```diff
@@ -37,6 +37,8 @@ selected_records = if cmd.opts.all?
                      end
                    end
 
+cmd.die "No such record." if selected_records.empty?
```
Comment (woodruffw) [inline lib/kbsecret/cli/kbsecret-login]: This change should go in a separate PR (along with the `env` change I mentioned in https://github.com/kbsecret/kbsecret/issues/36#issuecomment-338429767).

File: test/cli/test_login.rb [hunk 7]
```diff
@@ -0,0 +1,237 @@
+require_relative "test_helper"
```
Comment (woodruffw) [inline test/cli/test_login.rb]: Add `# frozen_string_literal: true` to the beginning of this file (it's good practice in Ruby 2.3+).

File: test/cli/test_login.rb (hunk 1 above)
Comment (woodruffw) [reply test/cli/test_login.rb]: Oh, and as a general note: you don't need to interpolate variables that are already strings.

These two lines are equivalent:

```ruby
type "#{foo}"
type foo
```

File: test/cli/test_login.rb [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    end
+
+    # retrieve logins:
+    run_command "kbsecret login -s #{session} -a" do |cmd|
+      cmd.wait
+      assert_match pattern, cmd.output
+    end
+  ensure
+    # remove session:
+    run_command_and_stop "kbsecret session rm -d #{session}"
```
Comment (woodruffw) [inline test/cli/test_login.rb]: Make sure you delete the login records here, as well.

File: test/cli/test_login.rb [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+  end
+
+  def test_all
+    session = "test-session"
+    label = "test-session-login"
+    label1 = "test-session-login1"
+    username = "user"
+    password = "pass"
+   # NOTE: need regex since we can't be sure of the order in which the logins will be retreived
+    pattern = /Label: (#{label}|#{label1})\n\tUsername: #{username}\n\tPassword: #{password}\nLabel: (#{label1}|#{label})\n\tUsername: #{username}\n\tPassword: #{password}\n/
```
Comment (woodruffw) [inline test/cli/test_login.rb]: Looking at this again, this sort of pattern will be difficult to update/maintain if the output changes. It might be better to take the output from the command and analyze each line individually instead (i.e., via a small state machine).

File: test/cli/test_login.rb (hunk 8 above)
Comment (dheles) [reply test/cli/test_login.rb]: unless something has changed, i believe that when you remove the session with -d, all of its records get destroyed

File: test/cli/test_login.rb (hunk 9 above)
Comment (dheles) [reply test/cli/test_login.rb]: maybe we can iterate on this one? i have a sizable backlog of additional tests to PR once PR #38 gets merged

File: test/cli/test_login.rb (hunk 8 above)
Comment (woodruffw) [reply test/cli/test_login.rb]: Yep, you're right. Wasn't thinking 😄

File: test/cli/test_login.rb (hunk 9 above)
Comment (woodruffw) [reply test/cli/test_login.rb]: That's fine -- just mark it with a `NOTE` or `TODO`.

File: test/cli/test_login.rb [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+    run_command "kbsecret login -s #{session} #{session_label}" do |cmd|
+      cmd.wait
+      assert_match error_pattern_no_session, cmd.stderr
+    end
+  ensure
+    # remove default login:
+    run_command_and_stop "kbsecret rm #{default_label}"
+
+    # fail to retreive default login:
+    run_command "kbsecret login -x #{default_label}" do |cmd|
```
Comment (woodruffw) [inline test/cli/test_login.rb]: This should probably be somewhere other than the `ensure` block. It looks like we've covered this part of the code pretty well, though, so it'd be okay to drop this test if you don't want to fit it somewhere else.

File: lib/kbsecret/cli/kbsecret-env [hunk 11]
```diff
@@ -37,6 +37,8 @@ selected_records = if cmd.opts.all?
                      end
```
Comment (woodruffw) [inline lib/kbsecret/cli/kbsecret-env]: This file needs to be dropped now that #38 is merged.

File: lib/kbsecret/cli/kbsecret-env [hunk 12]
```diff
@@ -37,6 +37,8 @@ selected_records = if cmd.opts.all?
...
                    end
```
Comment (woodruffw) [inline lib/kbsecret/cli/kbsecret-env]: Same with this change.
